crates/wash-lib/src/parser/mod.rs

```diff
@@ -985,23 +731,106 @@ pub fn get_config(opt_path: Option<PathBuf>, use_env: Option<bool>) -> Result<Pr
         })?
         .try_deserialize::<serde_json::Value>()?;
 
-    let raw_project_config: RawProjectConfig = serde_json::from_value(json_value)?;
+    let toml_project_config: WasmcloudDotToml = serde_json::from_value(json_value)?;
 
-    raw_project_config
+    toml_project_config
         .convert(project_path)
         .map_err(|e: anyhow::Error| anyhow!("{} in {}", e, wasmcloud_path.display()))
 }
 
-impl RawProjectConfig {
+/// The wasmcloud.toml specification format as de-serialization friendly project configuration data
+///
+/// This structure is normally directly de-serialized from `wasmcloud.toml`,
+/// and is used to build a more structured [`ProjectConfig`] object.
+///
+/// Below is an example of each option in the wasmcloud.toml file. A real example
+/// only needs to include the fields that are relevant to the project.
+///
+/// ```rust
+/// use wash_lib::parser::WasmcloudDotToml;
+///
+/// let component_toml = r#"
+/// language = "rust"
+/// type = "component"
+/// name = "testcomponent"
+/// version = "0.1.0"
+/// "#;
+/// let config: WasmcloudDotToml = toml::from_str(component_toml).expect("should deserialize");
+/// eprintln!("{config:?}");
+/// ```
+#[derive(Deserialize, Debug)]
+pub struct WasmcloudDotToml {
+    /// The language of the project, e.g. rust, tinygo. This is used to determine which config to parse.
+    pub language: String,
+
+    /// The type of project. This is a string that is used to determine which type of config to parse.
+    /// The toml file name is just "type" but is named project_type here to avoid clashing with the type keyword in Rust.
+    #[serde(rename = "type")]
+    pub project_type: String,
+
+    /// Name of the project. Optional if building a Rust project, as it can be inferred from Cargo.toml.
+    pub name: Option<String>,
+
+    /// Semantic version of the project. Optional if building a Rust project, as it can be inferred from Cargo.toml.
+    pub version: Option<Version>,
+
+    /// Monotonically increasing revision number.
+    #[serde(default)]
+    pub revision: i32,
+
+    /// Path to the directory where the project is located. Defaults to the current directory.
+    /// This path is where build commands will be run.
+    pub path: Option<PathBuf>,
+
+    /// Path to the directory where the WIT world and dependencies can be found. Defaults to a `wit`
+    /// directory in the project root.
+    pub wit: Option<PathBuf>,
+
+    /// Path to the directory where the built artifacts should be written. Defaults to a `build`
+    /// directory in the project root.
+    pub build: Option<PathBuf>,
+
+    /// Configuration relevant to components
+    #[serde(default)]
+    pub component: ComponentConfig,
+
+    /// Configuration relevant to providers
+    #[serde(default)]
+    pub provider: ProviderConfig,
```

Comment:
Nit: I feel like it makes more sense for each of these configs to be options rather than defaulted. I know most of them contain options internally, but for something that can (and often) be entirely omitted, an option makes more sense to me

Reply:
Yeah I considered that when doing some of the refactoring, and I landed on this because if we have the default here then all of the code can just deal with the default values, rather than many different places having to do a `provider.map(|p| p.some_sub_field).unwrap_or_default()` or a `if let Some(provider) = <>`

Open to changing this in the future, though, if it does get cleaner that way

Reply:
I know this is one you said would bother you more, so I'll let you resolve if you're satisfied w/ my explanation